proposals/20201026-creating-control-plane.md

```diff
@@ -0,0 +1,433 @@
+
+```go
+type NestedComponentSpec struct {
+    // ComponentSpec defines the common information for creating the component
+    // +optional
+    CommonSpec addonv1alpha1.CommonSpec `json:"commonSpec",omitempty`
+    
+    // Resources defines the amount of computing resources that will be used by this component
+    // +optional
+    Resources corev1.ResourceRequirements `json:"resources",omitempty`
+    
+    // Replicas defines the number of replicas in the component's workload 
+    // +optional
+    Replicas int32 `json:"replicas",omitempty`
+
+    // PatchSpecs includes the user specifed settings
+    // +optional
+    PatchSpec addonv1alpha1.PatchSpec `json:"patchSpec,omitempty"`
+}
+```
+
+The `CommonSpecs` and the `PatchSpec` are defined in [kubebuilder-declarative-pattern](https://github.com/kubernetes-sigs/kubebuilder-declarative-pattern/blob/1cbf859290cab81ae8e73fc5caebe792280175d1/pkg/patterns/addon/pkg/apis/v1alpha1/common_types.go):
+
+
+```go
+// CommonSpec defines the set of configuration attributes that must be exposed on all addons.
+type CommonSpec struct {
+	// Version specifies the exact addon version to be deployed, eg 1.2.3
+	// It should not be specified if Channel is specified
+	Version string `json:"version,omitempty"`
+	// Channel specifies a channel that can be used to resolve a specific addon, eg: stable
+	// It will be ignored if Version is specified
+	Channel string `json:"channel,omitempty"`
+}
+
+// +k8s:deepcopy-gen=true
+type PatchSpec struct {
+	Patches []*runtime.RawExtension `json:"patches,omitempty"`
+}
+```
+
+If the designated component's providers do not allow patching, the `PatchSpec` will be ignored. All Nested CRs need to contain a `NestedComponentSpec`. We will not discuss how NCPT and NCP controller generates the `NestedComponentSpec` in this proposal. However, a possible way is to add three fields to the NCPT CRD, and has each of them holds the `NestedComponentSpec` for one of the components. The NCPT will look like the following
+
+```go
+// NestedControlPlaneTemplateSpec defines the desired state of NestedControlPlaneTemplate
+type NestedControlPlaneTemplateSpec struct {
+    // other fields ...    
+    
+    // NestedEtcd contains information required by the etcd provider to create the Etcd
+    NestedEtcd NestedComponentSpec `json:"nestedEtcd"`
+
+    // NestedKAS contains information required by the apiserver provider to create the apiserver
+    NestedKAS NestedComponentSpec `json:"nestedKAS"`
+
+    // NestedKCM contains information required by the controller-manager provider to create the controller-manager
+    NestedKCM NestedComponentSpec `json:"nestedKCM"`
+}
+```
+
+We will let users specify component providers through NCP, implement a native mode that has each sub-controller to create the component on the super cluster, and define three new types `EtcdProvider`, `KASProvider`, and the `KCMProvider`:
```

Comment:
Could we run all controllers all the time instead of allowing to turn them on/off? If controllers aren't reconciling anything because there are no resources / events, it should be fine?

Reply:
The idea here would be that in the config (or flags 🤮 ) we can turn off specific controllers and if you wanted to use an external etcd cluster provider like `etcdadm` or `etcd-cluster-operator` you can choose to flip off the built-in provider and you just supply your own controller listening for `NEtcd` and creating w/e the necessary components for that implementation. This will change a bit of the implementation as per - https://github.com/kubernetes-sigs/cluster-api-provider-nested/pull/11/files/4b6d8aea98ed04eea543f2e29319567d4bab5207#r521166698